lib/Listener/CircleMembershipListener.php

```diff
@@ -0,0 +1,179 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Listener;
+
+use OCA\Circles\Events\AddingCircleMemberEvent;
+use OCA\Circles\Events\RemovingCircleMemberEvent;
+use OCA\Circles\Model\Member;
+use OCA\Talk\Exceptions\ParticipantNotFoundException;
+use OCA\Talk\Manager;
+use OCA\Talk\Model\Attendee;
+use OCA\Talk\Participant;
+use OCA\Talk\Service\ParticipantService;
+use OCP\App\IAppManager;
+use OCP\EventDispatcher\Event;
+use OCP\IGroupManager;
+use OCP\ISession;
+use OCP\IUser;
+use OCP\IUserManager;
+
+class CircleMembershipListener extends AMembershipListener {
+	/** @var ISession */
+	private $session;
+	/** @var IUserManager */
+	private $userManager;
+
+	public function __construct(Manager $manager,
+								IAppManager $appManager,
+								IGroupManager $groupManager,
+								ParticipantService $participantService,
+								IUserManager $userManager,
+								ISession $session) {
+		parent::__construct(
+			$manager,
+			$appManager,
+			$groupManager,
+			$participantService
+		);
+		$this->userManager = $userManager;
+		$this->session = $session;
+	}
+
+	public function handle(Event $event): void {
+		if ($event instanceof AddingCircleMemberEvent) {
+			$this->addingCircleMemberEvent($event);
+		}
+
+		if ($event instanceof RemovingCircleMemberEvent) {
+			$this->removeFormerMemberFromRooms($event);
+		}
+	}
+
+	protected function addingCircleMemberEvent(AddingCircleMemberEvent $event): void {
+		$roomsForTargetCircle = $this->manager->getRoomsForActor(Attendee::ACTOR_CIRCLES, $event->getCircle()->getSingleId());
+		$roomsToAdd = [];
+		foreach ($roomsForTargetCircle as $room) {
+			$roomsToAdd[$room->getId()] = $room;
+		}
+
+		// Check nested circles
+		$memberships = $event->getCircle()->getMemberships();
+		foreach ($memberships as $membership) {
+			$parentId = $membership->getCircleId();
+			$parentRooms = $this->manager->getRoomsForActor(Attendee::ACTOR_CIRCLES, $parentId);
+			foreach ($parentRooms as $room) {
+				if (isset($roomsToAdd[$room->getId()])) {
+					continue;
+				}
+				$roomsToAdd[$room->getId()] = $room;
+			}
+		}
+
+
+		if (empty($roomsToAdd)) {
+			// The circle is not in any room => bye!
+			return;
+		}
+
+		// This member is a "membership" in circles which links to entities such as users, groups or circles
+		$newMember = $event->getMember();
+		// Get the base circle of the membership
+		$basedOnCircle = $newMember->getBasedOn();
+		// Get all (nested) memberships in the added $newMember as a flat list
+		$userMembers = $basedOnCircle->getInheritedMembers();
+
+		$invitedBy = $newMember->getInvitedBy();
+		if ($invitedBy->getUserType() === Member::TYPE_USER && $invitedBy->getUserId() !== '') {
+			$this->session->set('talk-overwrite-actor', $invitedBy->getUserId());
+		} elseif ($invitedBy->getUserType() === Member::TYPE_APP && $invitedBy->getUserId() === 'occ') {
```

Comment:
while userId='occ' have no reason to change, `$invitedBy->getUserId() === 'occ'` can be replaced by `$invitedBy->getBasedOn()->getSource() === Member::APP_OCC`